crates/miden-objects/src/account/storage/map/witness.rs

```diff
@@ -1,50 +1,112 @@
+use alloc::collections::BTreeMap;
+
 use miden_crypto::merkle::{InnerNodeInfo, SmtProof};
 
 use crate::Word;
+use crate::account::StorageMap;
+use crate::errors::StorageMapError;
 
 /// A witness of an asset in a [`StorageMap`](super::StorageMap).
 ///
 /// It proves inclusion of a certain storage item in the map.
+///
+/// ## Guarantees
+///
+/// This type guarantees that the original key-value pairs it contains are all present in the
+/// contained SMT proof. Note that the inverse is not necessarily true. The proof may contain more
+/// entries than the map because to prove inclusion of a given original key A an
+/// [`SmtLeaf::Multiple`](miden_crypto::merkle::SmtLeaf::Multiple) may be present that contains both
+/// keys hash(A) and hash(B). However, B may not be present in the key-value pairs and this is a
+/// valid state.
 #[derive(Debug, Clone, PartialEq, Eq)]
-pub struct StorageMapWitness(SmtProof);
+pub struct StorageMapWitness {
+    proof: SmtProof,
+    map: BTreeMap<Word, Word>,
```

Comment:
Same as above comment: would `entries` be a better name here?

I would also maybe leave a doc comment that the `key` here is the un-hashed key.